cmd/subctl/recover_broker_info.go

```diff
@@ -0,0 +1,172 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subctl
+
+import (
+	"errors"
+	"fmt"
+	"github.com/submariner-io/submariner-operator/api/v1alpha1"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/subctl/internal/cli"
+	"github.com/submariner-io/subctl/internal/constants"
+	"github.com/submariner-io/subctl/internal/exit"
+	"github.com/submariner-io/subctl/internal/restconfig"
+	"github.com/submariner-io/subctl/pkg/broker"
+	"github.com/submariner-io/subctl/pkg/cluster"
+)
+
+var recoverRestConfigProducer = restconfig.NewProducer().WithPrefixedContext("broker").
```

Comment:
I find the whole UX of this confusing, since you have `--context`, `--namespace`, `--brokercontext` and `--brokernamespace`.

As a user I'm supposed to provide both `--context` and `--brokercontext` (unless the broker happens to be on the same cluster).
I would expect to provide, as a user, only the `--context` and the command should figure out whatever details (technically only components and custom domains right now) from that cluster's connection to the broker.

I was thinking the idea was to be able to recover the info given only the broker cluster, but that isn't possible and makes the whole idea of "just specify `--brokercontext`" not really usable.
Seems that we have to get details from an installed Submariner instance anyhow.

I think we should only have `--context` flag.
If submariner is installed on the cluster, the command can then figure out most of the details from the `Submariner` CR and get any additional info by connecting to the broker from these details and querying it directly.

This will be a much simpler and friendlier UX, and avoids any confusion or "invalid combinations" that can bet sent by the user.

Reply:
The idea is to provide a smart interaction for the user. If there is no context specified (neither `--context` nor `--brokercontext`), then you can still recover the file. But if the user wants to specify either the Broker and Submariner clusters, this implementation allows that and errors out if the specified contexts are not correct. So this implementation doesn't force user to specify any context and the code sends out appropriate messages if something is wrong.

Also, what you are suggesting is a design change. The right time for it would have been during the design phase.

Reply:
> The idea is to provide a smart interaction for the user. If there is no context specified (neither `--context` nor `--brokercontext`), then you can still recover the file. But if the user wants to specify either the Broker and Submariner clusters, this implementation allows that and errors out if the specified contexts are not correct. So this implementation doesn't force user to specify any context and the code sends out appropriate messages if something is wrong.

I don't understand your point.

The user must point (either explicitly or explicitly) to a cluster that has Submariner installed on it, right?
Why confuse them with a need to point to the cluster where the broker is, if the info is already available.

> 
> Also, what you are suggesting is a design change. The right time for it would have been during the design phase.

So you're saying that a worse design should be honored since it was agreed upon, instead of a better one?
That's not quite agile..

Reply:
It would be ideal if we could get all the info via the `Submariner` resource and not deal with a broker context. In order to do that we'd need to use the broker URL and credentials from the `Submariner` resource to retrieve the _submariner-k8s-broker-admin_  `Secret` and the Broker resource from the broker namespace.  But I'm not sure if the credentials have the permissions to do so. I assume that's why we're requiring a broker context from the user.

Reply:
What are those credentials? I can't seem to find it in the RBAC code to figure out if it has the correct permissions

Reply:
OK found it in https://github.com/submariner-io/subctl/blob/devel/pkg/broker/rbac.go#L51

So we could add a rule there to allow getting/listing brokers (which shouldn't be a problem as the broker doesn't have any secret data anyway, and the role is namespaced so the SA can only list the broker in the relevant namespace) and then it should be possible AFAICT.

Then we'd have a much simpler and less confusing UX

Reply:
I assume we'd have to build a `rest.Config` from  `Submariner.Spec.BrokerK8sApiServer`, `Submariner.Spec.BrokerK8sApiServerToken` et al but I **think** the associated RBAC is defined [here](https://github.com/submariner-io/subctl/blob/92115ef06fc50134c0f37cf3048afc6b64364911/pkg/broker/rbac.go#L82).

Reply:
At some point the token will be going away.

Reply:
I believe it's https://github.com/submariner-io/subctl/blob/92115ef06fc50134c0f37cf3048afc6b64364911/pkg/broker/rbac.go#L82 from which `Submariner.Spec.BrokerK8sApiServerToken` is derived. So in addition to getting the `Broker` resource, it would also need "get" access to `ServiceAccount` and `Secret`. This doesn't seem like an issue. @skitt WDYT?

Reply:
> At some point the token will be going away.

I assume the `Submariner` resource will still contain the info needed to connect to the broker cluster. Whatever form that is, I assume we should be able to use that here assuming we setup the proper RBAC.

Reply:
Yes, if joined clusters can connect to the broker, it would be possible to retrieve whatever information is needed to connect from somewhere else. (That may not always be true, but it is true for now.)

I’m under the impression we discussed this during the enhancement proposal review for the feature, but I can’t remember why we eschewed retrieving the broker credentials from a connected cluster instead of asking the user for broker credentials. Even if the user’s system can’t access the broker for some reason, the user could run `subctl` in a pod on a connected cluster.

Reply:
Seems the original discussion was happening here - https://github.com/submariner-io/enhancements/pull/144#discussion_r1050265381

But I think @tpantelis original assessment is either wrong, or I'm misunderstanding what we need to do.

To me it seems that the only access needed to be added to the default `Role: submariner-k8s-broker-cluster` is to be able to read a `Broker` resource (to get the `components` and `defaultCustomDomains` from it).

Am I missing something?

Reply:
Given a context capable of accessing the `submariner-operator` repository on a joined cluster, we can retrieve the broker secret named in the `Submariner` CR, and construct a client using that to access the broker using the appropriate credentials for that operator. The broker credentials don’t need to give access to a role which is itself capable of reading secrets.

Reply:
May I suggest the following compromise with regards to this PR? Given that there already is a working implementation using separate contexts, could we consider merging that (once other pending review comments have been resolved), and work on simplifying the UX separately?

Reply:
Generally I'm against this mentality of "lets merge this even if we don't agree and fix it later in a separate PR".

* We still need to review this "less optimal solution", which is harder to review IMHO because of all this confusion.
* It still gets written in the history forever, and in a segmented fashion which makes browsing the history a bit more complicated.
* There's a risk that later may not actually happen.

If this was some unrelated change, merging now and adding or fixing something else later makes more sense, but as it touches on the PR contents themselves  IMHO this isn't a good compromise.

But then again I'm not a project committer/owner so eventually its up to committers/owners to decide.

Reply:
Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.

I take it “a segmented fashion” refers to the use of multiple commits; that would be addressed in the review.

The risk that later may not actually happen also applies here — requesting changes may result in the changes being implemented, or it may result in the PR being abandoned; there’s no guarantee.

Reply:
> Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.

I already replied to this talking point here - https://github.com/submariner-io/subctl/pull/467#discussion_r1105879456

Reply:
> > Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.
> 
> I already replied to this talking point here - [#467 (comment)](https://github.com/submariner-io/subctl/pull/467#discussion_r1105879456)

No, you replied to a different point. I’m not saying this should be applied because it implements the proposed enhancement. I’m saying that the enhancement proposal makes this PR easy to understand, and that if changes are to be made, for consistency the EP should be updated too.

Reply:
The consensus is to finish this PR as is since it does follow the previously agreed upon proposal. Subsequently the feasibility of simplifying the UX by removing the `brokercontext` arg will be explored and the enhancement proposal modified accordingly followed by the implementation.